Stopped reconciliation if quorumloss annotation is set by operator. (#446)
gardener/etcd-druid: docs/operation/Recover_From_Etcd_Permanent_Quorum_Loss.md

## Quorum loss in ETCD Cluster

> [Quorum loss](https://etcd.io/docs/v3.4/op-guide/recovery/) means when majority of ETCD pods(Greater than or equal to n/2 + 1) are down simultaneously for some reason. 

Author: Quorum loss is referred there only

> [Quorum loss](https://etcd.io/docs/v3.4/op-guide/recovery/) means when majority of ETCD pods(Greater than or equal to n/2 + 1) are down simultaneously for some reason. 

Author: Done

There are two types of quorum loss that can happen to [ETCD multinode cluster](https://github.com/gardener/etcd-druid/tree/master/docs/proposals/multi-node) :

> 1. **Transient quorum loss** - A quorum loss is called transient when majority of ETCD pods are down simultaneously for some time. The pods may be down for network unavailability, high resource usages etc. When the pods come back after some time, they can re-join to the cluster and the quorum is recovered automatically without any manual intervention. The data directory must not be corrupted for majority of pods in this type of quorum loss.

Author: Done

> 2. **Permanent quorum loss** - A quorum loss is called permanent when majority of ETCD pods are down simultaneously due to unavailability of data volumes that are/were attached to the ETCD pods to hold the ETCD data. If majority of ETCD data volumes are unavailable **SIMULTANEOUSLY**, the etcd cluster is not going to recover automatically from the quorum loss. A human operator will now need to intervene and execute the following steps to start a fresh multi-node ETCD cluster.

Reviewer: ```suggestion
2. **Permanent quorum loss** - A quorum loss is called permanent when majority of ETCD cluster members experience permanent failure, whether due to hardware failure or disk corruption etc. then the etcd cluster is not going to recover automatically from the quorum loss. A human operator will now need to intervene and execute the following steps to recover the multi-node ETCD cluster.
```

Author: Done

      ```

      Change the `initial-cluster` field to have only one member (`etcd-main-0`) in the string. It should now look like this:

      ```
      # Initial cluster
        initial-cluster: etcd-main-0=https://etcd-main-0.etcd-main-peer.default.svc:2380
      ```
   6. Scale up the `etcd-main` statefulset replicas to `1`

      `kubectl scale sts etcd-main --replicas=1`
   7. Wait for the single-member etcd cluster to be completely ready.

      `kubectl get pods etcd-main-0` will give the following output when ready:
      ```
      NAME          READY   STATUS    RESTARTS   AGE
      etcd-main-0   2/2     Running   0          1m
      ```
   8. Add the following annotation to the `etcd-main` statefulset `kubectl annotate sts etcd-main gardener.cloud/scaled-to-multi-node="true"`
   9. Remove the following annotation from the `Etcd` resource `etcd-main`: `kubectl annotate etcd etcd-main druid.gardener.cloud/ignore-reconciliation-`
   10. Finally add the following annotation to the `Etcd` resource `etcd-main`: `kubectl annotate etcd etcd-main gardener.cloud/operation="reconcile"`

>    11. Verify that the etcd cluster is formed correctly.

Reviewer: I'm not getting how you are verifying whether the etcd cluster formed correctly or not.

Author: I made some changes for this. Please check

## Quorum loss in ETCD Cluster

> [Quorum loss](https://etcd.io/docs/v3.3/op-guide/recovery/) means when majority of ETCD pods(Greater than or equal to n/2 + 1) are down simultaneously for some reason. 

Reviewer 2: Please link the latest documentation, or atleast to etcd `v3.4`, since druid uses etcd `v3.4` version at the moment.

There are two types of quorum loss that can happen to ETCD multinode cluster -

1. **Transient quorum loss** - A quorum loss is called transient when majority of ETCD pods are down simultaneously for some time. The pods may be down for network unavailability, high resource usages etc. When the pods come back after some time, they can re-join to the cluster and the quorum is recovered automatically without any manual intervention. The data directory must not be corrupted for majority of pods in this type of quorum loss.

2. **Permanent quorum loss** - A quorum loss is called permanent when majority of ETCD pods are down simultaneously due to unavailability of data volumes that are/were attached to the ETCD pods to hold the ETCD data. If majority of ETCD data volumes are unavailable **SIMULTANEOUSLY**, the etcd cluster is not going to recover automatically from the quorum loss. A human operator will now need to intervene and execute the following steps to start a fresh multi-node ETCD cluster.

## ETCD clusters in shoot control plane

> If permanent quorum loss occurs to a [Multinode ETCD cluster](https://github.com/gardener/etcd-druid/tree/master/docs/proposals/multi-node), the operator needs to note down the PVCs, configmaps, statefulsets, CRs etc related to that ETCD cluster and work on those resources only. Following steps guide a human operator to recover from permanent quorum loss of a ETCD cluster. We assume the name of the ETCD CR for the ETCD cluster is `etcd-main`

Reviewer 2: Please add the multi-node proposal link to line 4 since that is the first reference of the term, and remove it from this line here.

>    1. Add the following annotation to the etcd object `k annotate etcd etcd-main druid.gardener.cloud/ignore-reconciliation="true"`

Reviewer 2: ```suggestion
   1. Add the following annotation to the `Etcd` resource `kubectl annotate etcd etcd-main druid.gardener.cloud/ignore-reconciliation="true"`
```

>    3. Note down the configmap name that is attached to the `etcd-main` statefulset. If you describe the statefulset with `k describe sts etcd-main`, look for the lines similar to following lines to identify attached configmap name. It will be needed at later stages:

Reviewer 2: ```suggestion
   3. Note down the configmap name that is attached to the `etcd-main` statefulset. If you describe the statefulset with `kubectl describe sts etcd-main`, look for the lines similar to following lines to identify attached configmap name. It will be needed at later stages:
```

   ```
     Volumes:
      etcd-config-file:
       Type:      ConfigMap (a volume populated by a ConfigMap)
       Name:      etcd-bootstrap-4785b0
       Optional:  false
   ```

>    3. Scale down the `etcd-main` statefulset replicas to 0

Reviewer 2: ```suggestion
   3. Scale down the `etcd-main` statefulset replicas to `0`
```

      `kubectl scale sts etcd-main --replicas=0`

>    4. The PVCs will look like the following on listing them with the command `k get pvc` :

Reviewer 2: ```suggestion
   4. The PVCs will look like the following on listing them with the command `kubectl get pvc` :
```

      
      ```
      main-etcd-etcd-main-0        Bound    pv-shoot-garden-aws-c1c6154e-6506-4017-9cfc-4f1cc9afa730   80Gi       RWO            gardener.cloud-fast   385d
      ```

>       Delete all PVCs that were attached to ETCD main cluster.

Reviewer 2: ```suggestion
      Delete all PVCs that are attached to `etcd-main` cluster.
```

>       `k delete pvc -l instance=etcd-main`

Reviewer 2: ```suggestion
      `kubectl delete pvc -l instance=etcd-main`
```

>    5. Edit the `etcd-main` configmap’s (Ex: etcd-bootstrap-4785b0).

Reviewer 2: ```suggestion
   5. Edit the `etcd-main` cluster's configmap (ex: `etcd-bootstrap-4785b0`) as follows:
```

      
      Find the `initial-cluster` field in the configmap. It will look like the following:
      ```
      # Initial cluster
        initial-cluster: etcd-main-0=https://etcd-main-0.etcd-main-peer.default.svc:2380,etcd-main-1=https://etcd-main-1.etcd-main-peer.default.svc:2380,etcd-main-2=https://etcd-main-2.etcd-main-peer.default.svc:2380
      ```

>       Change the `initial-cluster` field to have only one member (etcd-main-0) in the string to make it look like the following:

Reviewer 2: ```suggestion
      Change the `initial-cluster` field to have only one member (`etcd-main-0`) in the string. It should now look like this:
```

      ```
      # Initial cluster
        initial-cluster: etcd-main-0=https://etcd-main-0.etcd-main-peer.default.svc:2380
      ```

>    6. Scale up the `etcd-main` statefulset replicas to 1

Reviewer 2: ```suggestion
   6. Scale up the `etcd-main` statefulset replicas to `1`
```

      `kubectl scale sts etcd-main --replicas=1`

>    7. Wait for the single member etcd to be completely ready.

Reviewer 2: ```suggestion
   7. Wait for the single-member etcd cluster to be completely ready.
```

>       `k get pods etcd-main-0` will give the following output when ready:

Reviewer 2: ```suggestion
      `kubectl get pods etcd-main-0` will give the following output when ready:
```

      ```
      NAME          READY   STATUS    RESTARTS   AGE

>       etcd-main-0   2/2     Running   0          13h

Reviewer 2: ```suggestion
      etcd-main-0   2/2     Running   0          1m
```
When the sts is scaled up, the pod would have started up recently, rather than `13h` ago, right?

      ```

>    8. Remove the following annotation from the etcd object `k annotate etcd etcd-main druid.gardener.cloud/ignore-reconciliation-`

Reviewer 2: ```suggestion
   8. Remove the following annotation from the `Etcd` resource `etcd-main`: `kubectl annotate etcd etcd-main druid.gardener.cloud/ignore-reconciliation-`
```

>    9. Add the following annotation to the `etcd-main` statefulset `k annotate sts etcd-main gardener.cloud/scaled-to-multi-node="true"`

Reviewer 2: ```suggestion
   9. Add the following annotation to the `etcd-main` statefulset `kubectl annotate sts etcd-main gardener.cloud/scaled-to-multi-node="true"`
```

>    10. Finally add the following annotation to the etcd object `k annotate etcd etcd-main gardener.cloud/operation="reconcile"`

Reviewer 2: ```suggestion
   10. Finally add the following annotation to the `Etcd` resource `etcd-main`: `kubectl annotate etcd etcd-main gardener.cloud/operation="reconcile"`
```

>    11. Verify that the cluster is formed correctly.

Reviewer 2: ```suggestion
   11. Verify that the etcd cluster is formed correctly.
```

       All the `etcd-main` pods will have outputs similar to following:

>        ```
>        NAME          READY   STATUS    RESTARTS   AGE
>        etcd-main-0   2/2     Running   0          13h
>        etcd-main-1   2/2     Running   0          13h
>        etcd-main-2   2/2     Running   0          13h
>        ```

Reviewer 2: ```suggestion
       ```
       NAME          READY   STATUS    RESTARTS   AGE
       etcd-main-0   2/2     Running   0          5m
       etcd-main-1   2/2     Running   0          1m
       etcd-main-2   2/2     Running   0          1m
       ```
```

>        Additionally, the operator can get into etcd container of any of the ETCD main pod with `k exec -it etcd-main-0 -c etcd /bin/sh` and execute the command `etcdctl --cacert=/var/etcd/ssl/client/ca/bundle.crt --cert=/var/etcd/ssl/client/client/tls.crt --key=/var/etcd/ssl/client/client/tls.key --endpoints=https://etcd-main-local:2379

Reviewer 2: ```suggestion
       Additionally, the operator can get a shell into the `etcd` container of any of the etcd-main pods, with `kubectl exec -it etcd-main-0 -c etcd /bin/sh` and execute the command `etcdctl --cacert=/var/etcd/ssl/client/ca/bundle.crt --cert=/var/etcd/ssl/client/client/tls.crt --key=/var/etcd/ssl/client/client/tls.key --endpoints=https://etcd-main-local:2379
```

>  endpoint status --cluster -w table` and check the output similar to following to find if all the ETCD main members are listed properly or not.

Reviewer 2: ETCD main -> `etcd-main`

       |                      ENDPOINT                       |        ID        | VERSION | DB SIZE | IS LEADER | IS LEARNER | RAFT TERM | RAFT INDEX | RAFT APPLIED INDEX | ERRORS |
       +-----------------------------------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
       | https://etcd-main-2.etcd-main-peer.default.svc:2379 | 127372b4166af24b |  3.4.13 |   25 kB |      true |      false |         9 |        213 |                213 |        |
       | https://etcd-main-1.etcd-main-peer.default.svc:2379 | 4e5250fc3982331a |  3.4.13 |   25 kB |     false |      false |         9 |        213 |                213 |        |
       | https://etcd-main-0.etcd-main-peer.default.svc:2379 | 8e9e05c52164694d |  3.4.13 |   25 kB |     false |      false |         9 |        213 |                213 |        |
       +-----------------------------------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
       ```

> ETCD cluster started this way will actually restore old data from the backup but there might be few minutes of data loss.

Reviewer 2: Can you please add this as a disclaimer to the starting of the document?

2. **Permanent quorum loss** - A quorum loss is called permanent when majority of ETCD pods are down simultaneously due to unavailability of data volumes that are/were attached to the ETCD pods to hold the ETCD data. If majority of ETCD data volumes are unavailable **SIMULTANEOUSLY**, the etcd cluster is not going to recover automatically from the quorum loss. A human operator will now need to intervene and execute the following steps to start a fresh multi-node ETCD cluster.

## ETCD clusters in shoot control plane
If permanent quorum loss occurs to a multinode ETCD cluster, the operator needs to note down the PVCs, configmaps, statefulsets, CRs etc related to that ETCD cluster and work on those resources only. Following steps guide a human operator to recover from permanent quorum loss of a ETCD cluster. We assume the name of the ETCD CR for the ETCD cluster is `etcd-main`.

**Note:** Please note that manually restoring etcd can result in data loss. This guide is the last resort to bring an ETCD cluster up and running again.

   1. Add the following annotation to the `Etcd` resource `kubectl annotate etcd etcd-main druid.gardener.cloud/ignore-reconciliation="true"`

>    3. Note down the configmap name that is attached to the `etcd-main` statefulset. If you describe the statefulset with `kubectl describe sts etcd-main`, look for the lines similar to following lines to identify attached configmap name. It will be needed at later stages:

Reviewer 3: ```suggestion
   2. Note down the configmap name that is attached to the `etcd-main` statefulset. If you describe the statefulset with `kubectl describe sts etcd-main`, look for the lines similar to following lines to identify attached configmap name. It will be needed at later stages:
```

Author: Done

      
      Find the `initial-cluster` field in the configmap. It will look like the following:
      ```
      # Initial cluster
        initial-cluster: etcd-main-0=https://etcd-main-0.etcd-main-peer.default.svc:2380,etcd-main-1=https://etcd-main-1.etcd-main-peer.default.svc:2380,etcd-main-2=https://etcd-main-2.etcd-main-peer.default.svc:2380
      ```

      Change the `initial-cluster` field to have only one member (`etcd-main-0`) in the string. It should now look like this:

      ```
      # Initial cluster
        initial-cluster: etcd-main-0=https://etcd-main-0.etcd-main-peer.default.svc:2380
      ```
   6. Scale up the `etcd-main` statefulset replicas to `1`

      `kubectl scale sts etcd-main --replicas=1`
   7. Wait for the single-member etcd cluster to be completely ready.

      `kubectl get pods etcd-main-0` will give the following output when ready:
      ```
      NAME          READY   STATUS    RESTARTS   AGE
      etcd-main-0   2/2     Running   0          1m
      ```
   8. Remove the following annotation from the `Etcd` resource `etcd-main`: `kubectl annotate etcd etcd-main druid.gardener.cloud/ignore-reconciliation-`

>    9. Add the following annotation to the `etcd-main` statefulset `kubectl annotate sts etcd-main gardener.cloud/scaled-to-multi-node="true"`

Reviewer 3: Can you reverse these two steps?
We should add the `scaled-to-multi-node` annotation to the sts before removing the `ignore-reconciliation` annotation from the etcd resource

Author: Done

There are two types of quorum loss that can happen to ETCD multinode cluster -

1. **Transient quorum loss** - A quorum loss is called transient when majority of ETCD pods are down simultaneously for some time. The pods may be down for network unavailability, high resource usages etc. When the pods come back after some time, they can re-join to the cluster and the quorum is recovered automatically without any manual intervention. The data directory must not be corrupted for majority of pods in this type of quorum loss.

2. **Permanent quorum loss** - A quorum loss is called permanent when majority of ETCD pods are down simultaneously due to unavailability of data volumes that are/were attached to the ETCD pods to hold the ETCD data. If majority of ETCD data volumes are unavailable **SIMULTANEOUSLY**, the etcd cluster is not going to recover automatically from the quorum loss. A human operator will now need to intervene and execute the following steps to start a fresh multi-node ETCD cluster.

## ETCD clusters in shoot control plane

> There are two [ETCD clusters](https://github.com/gardener/etcd-druid/tree/master/docs/proposals/multi-node) running in shoot control plane. One is named as ETCD Events and another is named as ETCD main. The operator needs to take care of permanent quorum loss to a specific cluster. If permanent quorum loss occurs to ETCD events cluster, the operator needs to note down the PVCs, configmaps, statefulsets, CRs etc related to ETCD events cluster and work on those resources only. Following steps guide a human operator to recover from permanent quorum loss of **ETCD main cluster**. 

Reviewer 2: Please make the playbook generic enough for any `etcd-druid` user to use. Gardener uses `etcd-main` and `etcd-events`, but a community member using `etcd-druid` by itself can create ETCDs with any name like `etcd-test` or `my-etcd`. Please remove the gardener-specific details from this doc.

If permanent quorum loss occurs to a multinode ETCD cluster, the operator needs to note down the PVCs, configmaps, statefulsets, CRs etc related to that ETCD cluster and work on those resources only. Following steps guide a human operator to recover from permanent quorum loss of a ETCD cluster. We assume the name of the ETCD CR for the ETCD cluster is `etcd-main`.

**ETCD cluster in shoot control plane of gardener deployment:**
There are two [ETCD clusters](https://github.com/gardener/etcd-druid/tree/master/docs/proposals/multi-node) running in shoot control plane. One is named as `etcd-events` and another is named `etcd-main`. The operator needs to take care of permanent quorum loss to a specific cluster. If permanent quorum loss occurs to `etcd-events` cluster, the operator needs to note down the PVCs, configmaps, statefulsets, CRs etc related to `etcd-events` cluster and work on those resources only. 

> **Note:** Please note that manually restoring etcd can result in data loss. This guide is the last resort to bring an ETCD cluster up and running again.

Reviewer 2: ```suggestion
:warning: **Note:** Please note that manually restoring etcd can result in data loss. This guide is the last resort to bring an ETCD cluster up and running again.
```

Author: Done

   If etcd-druid and etcd-backup-restore is being used with gardener, then 

>    Go to control plane of affected shoot cluster.
> 
>    Use `gardenctl` to target the control plane of the affected shoot cluster. You can get the details to target the control plane from the Access tile in the shoot cluster dashboard.

Reviewer 2: ```suggestion
   Target the control plane of affected shoot cluster via `kubectl`. Alternatively, you can use [gardenctl](https://github.com/gardener/gardenctl-v2) to target the control plane of the affected shoot cluster. You can get the details to target the control plane from the Access tile in the shoot cluster details page on the Gardener dashboard. Ensure that you are targeting the correct namespace.
```

Author: Done
